File: 19.cpp
#include <iostream>

using namespace std;


bool isLeapYear(int year) {
    if (year % 100 == 0 && year % 400 != 0)
        return false;
    return (year % 4 == 0);
}


bool sundayStart(int days, bool leapYear) {
    /* Checks if month starts on a Sunday
     *  31 ends on Tuesday
     *  30 "    "  Monday
     *  28 "    "  Saturday
     */


    int lastDay = days % 7;

    if (days == 31)
        return lastDay == 3;
    else if (days == 30)
        return lastDay == 2;
    else if (days == 28) {
        if (!leapYear)
            return lastDay == 7;
        else
            return lastDay == 1;
    }

    return false;
}


void countSundays(int& sunday_count, int& year, int endYear, int months[], int overflow = 2) {
    /* January of 1901 starts on Tuesday; has an 'overflow' of 2 */

    // Add overflow to January
    months[0] = months[0] + overflow;

    while (year <= endYear) {
        for (int month = 0; month < 12; month++) {

            // sundayStart() and isLeapYear() are bool types and contain no loops
            if (sundayStart(months[month], isLeapYear(year)))
                sunday_count++;

            overflow = months[month] % 7;

            if (month < 11)
                months[month+1] = months[month+1] + overflow;
            else {// if end of year, add overflow to Jan of next year
                year++;
                countSundays(sunday_count, year, endYear, months, overflow);
            }
        }
    }
}


int main() {

    int startYear = 1901,
        endYear = 2000,
        sundays = 0,
        Jan = 31,
        Feb = 28,
        Mar = 31,
        Apr = 30,
        May = 31,
        June = 30,
        July = 31,
        Aug = 31,
        Sept = 30,
        Oct = 31,
        Nov = 30,
        Dec = 31;

    int months[] = {
        Jan,
        Feb,
        Mar,
        Apr,
        May,
        June,
        July,
        Aug,
        Sept,
        Oct,
        Nov,
        Dec
        };

    countSundays(sundays, startYear, endYear, months);

    cout << "Answer found: " << sundays << endl;

    return 0;
}

File: 19.py
#!/bin/python


def isLeapYear(year):
    if (not year % 100) and year % 400:
        return False
    return not year % 4


def countSundays():

    counter = 0 # Counts the months that start on a sunday

    for month in xrange(12):
        day = start_day[month]
        for year in xrange(1901, 2001):
            if day == 0:
                counter += 1
#                print "Year: ", year, "Month: ", month
            if isLeapYear(year):
                if day >= 6:
                    day = 0
                else:
                    day += 1

            if day < 6:
                day += 1
            else:
                day = 0
            
    return counter


if __name__ == '__main__':

    # Day of week each month starts in 1901 respectively
    start_day = [2, 5, 5, 1, 3, 6, 1, 4, 0, 2, 5, 0]

    print "Answer found: ", countSundays()
